### ISSUE TITLE
Multi-stage tuning still recompiles per stage, and AlgorithmResult misreports its hyperparameters

### ISSUE THREAD

**Issue by leon-k-martin** [virtual-twin/tvbo#91]
Found by a review of `dev` (`1e0b66dd..63aaec4e`, the compile-once work) and verified against the templates.

## The recompile still happens for the schedule shape the schema documents

`tvbo-tvboptim-algorithm.py.mako:804` marks these static:

```python
_core_statics = ["model_fn", "n_iterations", "print_every", "save_every", "verbose"]
```

and the stage driver passes a per-stage value:

```python
# tvbo-tvboptim-experiment.py.mako:4183
n_iterations=_sd['n_iterations'],
```

`save_every` and `print_every` are not passed at all, so both default to `_smart_interval(n_iterations)` — meaning **three statics change together** whenever a stage varies `n_iterations`. That is a fresh jit cache entry and a full XLA compile per stage: the 47-minute problem this change targets.

`AlgorithmStage` documents exactly that shape (`schema/tvbo_datamodel.yaml:3986`: *"Each stage overrides n_iterations + selected hyperparameters"*).

**The new guard does not catch it.** `test_multistage_tuning_core_compiles_once` builds both stages with the same `n_iter`, so it passes while the shipped schedule shape stays broken. Worth extending it to a varying-`n_iterations` schedule — it should fail today.

## `AlgorithmResult.hyperparameters` reports values that were never used

```mako
_hyperparams = Bunch(
% for pname, pval in hyperparam_dict.items():
    ${pname}=${pval},
% endfor
)
```

These are the **codegen literal defaults**, not the traced per-stage arguments the run actually consumed. With a schedule passing `eta=0.10` then `eta=0.05`, every stage's `AlgorithmResult` reports the literal `eta=0.5`. Anything reading `result.hyperparameters` for provenance — reports, `estimate__` sourcing — records a value that was never used. (See also the `estimate__` last-writer note in the existing tuning work.)

## Two smaller ones from the same move

* The `if verbose: logger.info("Running <algo> for N iterations...")` header moved inside the jitted core, so it now fires at **trace** time — once per compile, not once per stage. With compile-once working, stages 2..N print no header. Any host-side statement later added in that region will silently run only on the first stage.
* `_rec_positions` / `_rec_idx` are computed twice from the same inputs — once in `run_<algo>` (line 579) and identically inside the core (line 984). `run_` uses `_rec_idx` to subsample the scan `ys`; the core uses `_n_records` to size the `__rec` buffers. If either changes alone they disagree and `result_history` mixes differently-subsampled traces, with no error.

Also dead after the move: the list-based `result_history = Bunch(...)` at line 390 and the `_<obs>_samples = []` lists at 417, both unconditionally superseded by the scan-ys rebuild at 653.

## Note for the 1.0 prose gate

`tvbo-tvboptim-algorithm.py.mako:589` (inside `_canon`) and `tvbo-tvboptim-experiment.py.mako:3459` carry 3-line and 6-line stacked `#` blocks. Those will fail `scripts/check_prose.py` once #88 lands and the gate goes blocking; both read as docstring material.

**Comment by leon-k-martin** [comment]
**Status update — triaged against `dev` at `636cb868`** (after #101, #98, #94, #88 merged).

**Verdict: valid**

Every code claim in the issue still reproduces verbatim on dev (636cb868): `n_iterations` is still a jit static while the stage driver still varies it per stage, `_hyperparams` is still built from codegen literals rather than the traced per-stage arguments, the verbose header still sits inside the jitted core, `_rec_positions` is still computed twice, and the list-based `result_history` / `_<obs>_samples` are still dead. I confirmed this empirically by rendering a 2-stage fic_eib schedule with n_iterations 3 then 7: the emitted static_argnames contain n_iterations/print_every/save_every and the stage call site passes only n_iterations, and the emitted `_hyperparams` reports eta=0.005 / window_size=150 while the schedule actually ran eta 0.1→0.05 and window_size 4→8. The only sub-item that has moved is the closing prose-gate note: #98 shipped the gate diff-scoped for .mako (whole-file mode is .py-only and CI runs it with `|| true`), so the pre-existing stacked blocks it predicted would fail do not fail. Since #101/#98/#88/#94 touched the algorithm template only to add the NaN guard and the update_every gate, nothing here was fixed by those merges.

Evidence:
- tvbo/templates/tvboptim/tvbo-tvboptim-algorithm.py.mako:834 (dev) — `_core_statics = ["model_fn", "n_iterations", "print_every", "save_every", "verbose"]`, consumed at :1515-1516 `_${algo_name}_tuning_core = jax.jit(..., static_argnames=(${_core_static_tuple}))`
- tvbo/templates/tvboptim/tvbo-tvboptim-experiment.py.mako:4316 (dev) — the `has_stages` branch passes `n_iterations=_sd['n_iterations']` and NO print_every/save_every; contrast the single-run branch at :4369-4371 which does pass them. Stage dicts are built at :4161 `_sd = {'n_iterations': int(_st.n_iterations), ...}`
- tvbo/templates/tvboptim/tvbo-tvboptim-algorithm.py.mako:380-383 — `print_every`/`save_every` fall back to `_smart_interval(n_iterations)`, so they are derived from the per-stage value; the core call at :626-629 forwards all four as statics
- Rendered proof (2-stage schedule, n_iterations 3 then 7, from tvbo/database/experiments/EI_Tuning_FIC_EIB_Optimization.yaml): /private/tmp/claude-501/-Users-leonmartin-bih-tools-tvbo/52d50406-e0f5-4c46-80d1-b5f85bd7ecbf/scratchpad/rendered91.py:2246-2255 emits `static_argnames=("model_fn", "n_iterations", "print_every", "save_every", "verbose", "use_ring")`; :2743-2758 emits `_stage_defs` with n_iterations 3 and 7; :2784 emits `n_iterations=_sd["n_iterations"]` with no print_every/save_every -> two distinct static tuples -> two XLA compiles
- tests/test_fic_eib_compile_once.py:22-34 — `_multistage_experiment(n_iter=3)` builds BOTH AlgorithmStages with the same `n_iterations=n_iter`, so `test_multistage_tuning_core_compiles_once` (tests/test_fic_eib_compile_once.py:108, @pytest.mark.slow) exercises only the shape that already compiles once. No test on dev builds a varying-n_iterations schedule.
- schema/tvbo_datamodel.yaml:3960 `AlgorithmStage`, with `n_iterations` required per stage (:3973) and the class description 'each stage overriding n_iterations and selected hyperparameters' — the shipped schema documents exactly the shape that recompiles

**Outstanding:** 1. Make the compile-once guard actually guard: change `_multistage_experiment` in tests/test_fic_eib_compile_once.py to give the two stages DIFFERENT `n_iterations` (e.g. 3 and 7). `test_multistage_tuning_core_compiles_once` should then fail with 2 compiles today. 2. Get `n_iterations` out of `_core_statics` (tvbo-tvboptim-algorithm.py.mako:834). This is structural, not a one-liner: the scan length, `_rec_positions`/`_n_records` and the `__rec` buffer shapes are all derived from it at Python level (:579-580, :1014-1016). The pattern already used for a varying window applies — pad to the schedule's max (`_max_window` / `use_ring` at experiment mako:4209 and algorithm mako:1023) and drive the scan with a traced iteration count plus a mask, so `n_iterations` becomes traced and the record buffers are sized once at max. `print_every`/`save_every` then need to be passed explicitly by the stage driver (or computed from the padded max) rather than falling out of `_smart_interval(n_iterations)` (:380-383). 3. Fix the misreport: emit `${pname}=${pname}` instead of `${pname}=${pval}` at algorithm mako:805-807 — the names are already parameters of `run_${algo_name}` (:341-342) — and add a test asserting `AlgorithmResult.hyperparameters.eta` equals the stage's eta, not the YAML default. Nothing currently asserts on `.hyperparameters` anywhere in tests/. 4. Move the `if verbose: logger.info("Running ... iterations")` header (:1004-1005) out of `_${algo_name}_tuning_core_impl` and into `run_${algo_name}` before the core call, so it fires once per stage rather than once per trace. 5. Compute `_rec_positions`/`_rec_idx`/`_n_records` once and pass them in, so `run_`'s subsampling and the core's buffer sizing cannot silently diverge. 6. Delete the dead `result_history = Bunch(...)` at :390-402 and `_${obs}_samples = []` at :417.

**PRs:** No PR.


<sub>Triaged by re-running the issue against current `dev`. Every “already fixed” claim was independently re-checked by a second pass before being recorded here; three were overturned.</sub>
